### ISSUE TITLE
Odd issue with certain float values

### ISSUE THREAD

**Issue by rsivek** [turbo/js#20]
I was running some tests on my kernel code when I started getting some very bizzare outputs. After some investigation it appears some values are not transferring correctly between JavaScript and the GPU.

```
// Allocate some data
var sampleSize = 128;
var testData = turbojs.alloc(sampleSize);

// Apparently WebGL doesn't like this number...
testData.data[0] = 9.12700582146515e-39;

// Do a simple reflection of the input values
console.log('Before:', testData.data);
turbojs.run(testData, `void main(void) {
commit(read());
}`);
console.log('After:', testData.data);
```

This gives me the following output in Chrome:
```
Before:
Float32Array(1024) [9.12700582146515e-39, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0…]
After:
Float32Array(1024) [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0…]
```

If I do some more normal number like 32 it works just fine:

```
// Allocate some data
var sampleSize = 128;
var testData = turbojs.alloc(sampleSize);

// It likes this number just fine
testData.data[0] = 32;

// Do a simple reflection of the input values
console.log('Before:', testData.data);
turbojs.run(testData, `void main(void) {
commit(read());
}`);
console.log('After:', testData.data);
```

This gives me the following output in Chrome:
```
Before:
Float32Array(1024) [32, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0…]
After:
Float32Array(1024) [32, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0…]
```

I had my colleague run the code on his machine, just to rule out the possibility that my graphics card has odd quirks, but his displayed the same behavior. 

Any idea about what's going on here?

**Comment by turbo** [comment]
@nanowizard,

Thanks for reporting. I will look into this.

Cheers.

**Comment by turbo** [comment]
@nanowizard,

That's not surprising. The GPU is dealing with 32bit floats, which have these limits: 

```text
FLT_MIN      = 1.175494e-38
FLT_MAX      = 3.402823e+38
```

`9.12700582146515e-39` is well below the limit. This value is nulled before it ever hits the GPU, probably when it's mapped by the turbo.js allocator.

**Comment by rsivek** [comment]
Great catch! I didn't even think to look at that. The way I actually got that number was I was trying to encode ascii 'abc' in the float input. Basically I'm doing something like this:

```
var test = new Float32Array(1);
var u8test = new Uint8Array(test.buffer);
u8test[0] = 0x61;
u8test[1] = 0x62;
u8test[2] = 0x63;
```
In JavaScript the float is represented as  `9.12700582146515e-39` in Little Endian. Since the browser didn't complain about it I figured it was fine. As you mentioned, however, it's outside the valid 32bit float range. In fact, I believe this is what's called a [denormalized or subnormal number](https://en.wikipedia.org/wiki/Denormal_number), which, it appears, the webgl context flushes to zero.

I suppose then the only way to overcome this would be modifying the turbo source to send in 32bit integers instead of floats?
